src/physics.cpp

```diff
@@ -1203,64 +1203,62 @@ void sample_secondary_photons(Particle& p, int i_nuclide)
   }
 }
 
-// split a particle on the basis of the statistical weight
-// of the current location
+// split a particle based on the statistical weight of the current location
 void split_particle(Particle& p)
 {
   // skip dead or no energy
   if (p.E() <= 0 || !p.alive()) return;
 
   bool in_domain = false;
   // todo this is a linear search - should do something more clever
-  ParticleWeightParams params;
-  for (const auto& domain : variance_reduction::ww_domains) {
-    in_domain = domain->find_params(p, params);
+  for (const auto& domain : variance_reduction::weight_window_domains) {
+    in_domain = domain->get_weight_window(p);
     if (in_domain)
       break;
   }
-
   // particle is not in any of the ww domains, do nothing
   if (!in_domain) return;
 
   // get the paramters
   double weight = p.wgt();
+  p.wgt_last() = weight;
+  const auto& weight_window = p.weight_window();
 
   // first check to see if particle should be killed for
   // weight cutoff
-  if ( p.wgt() < params.weight_cutoff ) {
+  if (p.wgt() < weight_window.weight_cutoff) {
     p.alive() = false;
     p.wgt() = 0.0;
-    p.wgt_last() = p.wgt();
     return;
   }
 
   // if particles weight is above the weight window
   // split until they are within the window
-  if (weight > params.upper_weight) {
-    double n_split = weight / params.upper_weight;
-    n_split = std::min(std::ceil(n_split), double(params.max_split));
+  if (weight > weight_window.upper_weight) {
+    double n_split = weight / weight_window.upper_weight;
+    n_split = std::min(
+      std::ceil(n_split), static_cast<double>(weight_window.max_split));
 
-    // possibility of some round off
-    for (int l = 0; l < int(n_split) - 1; l++) {
+    int i_split = std::round(n_split);
```

Comment:
the right thing to do here, isn't round, its to take nsplit, find the whole number component, then do the loop nsplit (int) times, but also sample a random number - which the probility of doing the floating point remainder, make sense?

Reply:
Just to check, this is equivalent to what was there before right? You're suggesting an improvement to that?

I agree about the remainder. I have a branch that assigns the remainder to the current particle and adds a secondary if needed. The current particle will be rouletted in this same function if its weight is below the weight window. Seems like it would avoid some additional particle splitting down the line. pshriwise@6593f81

Reply:
Maybe one clarifying piece of info would help -- I'm only using `std::round` here to get an int from the float, `n_split`. At this point `n_split` is already pinned to an integer value on line 1239 above. 

Maybe a better way of doing this would be `static_cast<int>` though?

Reply:
Actually I have an approach for handling the particle splitting I like even better: https://github.com/pshriwise/openmc/blob/39bb4a42f84dd9895fb9ca73eb3190e06b7e7e4d/src/physics.cpp#L1237-L1254